src/actions/NamespaceAction.ts

```diff
@@ -0,0 +1,64 @@
+import * as Api from '../services/Api';
+
+export enum NamespaceActionType {
+  RELOAD = 'reload',
+  API_INITIATE_REQUEST = 'api_initiate_request',
+  API_RECEIVE_LIST = 'api_receive_list'
+}
+
+export const reload = () => {
+  return {
+    type: NamespaceActionType.RELOAD
+  };
+};
+
+export const apiInitiateRequest = () => {
+  return {
+    type: NamespaceActionType.API_INITIATE_REQUEST
+  };
+};
+
+export const apiReceiveList = newList => {
+  return {
+    type: NamespaceActionType.API_RECEIVE_LIST,
+    list: newList,
+    receivedAt: Date.now()
+  };
+};
+
+export const asyncFetchNamespaces = () => {
+  return dispatch => {
+    dispatch(apiInitiateRequest());
+    return Api.GetNamespaces()
+      .then(response => response['data'])
+      .then(data => dispatch(apiReceiveList(data)));
+  };
+};
+
+const shouldFetchNamespaces = state => {
+  if (!state) {
+    return true;
+  } else if (state.namespaces.isFetching) {
+    return false;
+  } else {
+    return true;
+  }
+};
+
+export const fetchNamespacesIfNeeded = () => {
+  // Note that the function also receives getState()
+  // which lets you choose what to dispatch next.
+  // This is useful for avoiding a network request if
+  // a cached value is already available.
+  return (dispatch, getState) => {
+    console.log('fetchNamespacesIfNeeded()');
+    if (shouldFetchNamespaces(getState())) {
+      console.log('begin async fetching namespaces');
+      // Dispatch a thunk from thunk!
+      return dispatch(asyncFetchNamespaces());
+    } else {
+      // Let the calling code know there's nothing to wait for.
+      return Promise.resolve();
```

Comment:
It will return a resolved promise if we are fetching?
I suppose this function won't be used to actually get the namespaces, right?

Reply:
Yes it will return the resolved promise.  `fetchNamespacesIfNeeded()` action is mapped to the `refresh` callback:

https://github.com/vnugent/kiali-ui/blob/aeaf8358e654398a67c43e87cfad6572a9f5789b/src/containers/AutoUpdateNamespaceList.tsx#L14

As you see in the NamespaceDropdown.componentDidMount() and onSelect(), when call this.props.refresh(), Redux dispatches `fetchNamespacesIfNeeded()`

https://github.com/vnugent/kiali-ui/blob/aeaf8358e654398a67c43e87cfad6572a9f5789b/src/components/NamespaceDropdown.tsx#L19